src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -840,6 +858,106 @@ class ExtentPlacementManager {
+        DEFAULT,  // generation incremented with each rewrite. Extents will
+                  // be evicted when generation reaches MIN_COLD_GENERATION.
+        FAST,     // map all generations located in
+                  // [MIN_REWRITE_GENERATION, MIN_COLD_GENERATIOIN) to
+                  // MIN_COLD_GENERATION.
+      };
+
+      eviction_mode_t eviction_mode;
+      double stop_evict_ratio;
+      double default_evict_ratio;
+      double fast_evict_ratio;
+
+      bool should_do_fast_eviction(double alive_ratio) const {
+        return alive_ratio > fast_evict_ratio ||
+          eviction_mode == eviction_mode_t::FAST;
+      }
+
+      rewrite_gen_t adjust_generation_with_eviction(rewrite_gen_t gen) {
+        assert(is_rewrite_generation(gen));
+        rewrite_gen_t ret = gen;
+        switch(eviction_mode) {
+        case eviction_mode_t::STOP:
+          if (gen == MIN_COLD_GENERATION) {
+            ret = MIN_COLD_GENERATION - 1;
+          }
+          break;
+        case eviction_mode_t::DEFAULT:
+          break;
+        case eviction_mode_t::FAST:
+          if (gen >= MIN_REWRITE_GENERATION && gen < MIN_COLD_GENERATION) {
+            ret = MIN_COLD_GENERATION;
+          }
+          break;
+        default:
+          ceph_abort("impossible");
+        }
+        return ret;
+      }
+
+      // We change the state of eviction_mode according to the alive ratio
+      // of the hot tier.
+      //
+      // Use A, B, C, D to represent the state of alive ratio:
+      //   A: alive ratio <= stop_evict_ratio
+      //   B: alive ratio <= default_evict_ratio
+      //   C: alive ratio <= fast_evict_ratio
+      //   D: alive ratio >  fast_evict_ratio
+      //
+      // and use X, Y, Z to shorten the state of eviction_mode_t:
+      //   X: STOP
+      //   Y: DEFAULT
+      //   Z: FAST
+      //
+      // Then we can use a form like (A && X) to describe the current state
+      // of the hot tier, which indicates the alive ratio is less than or
+      // equal to stop_evict_ratio and current eviction mode is STOP.
+      //
+      // all valid state transitions show as follow:
+      //   (A && X) => (B && X) => (C && Y) => (D && Z) =>
+      //   (C && Z) => (B && Y) => (A && X)
+      void maybe_update_eviction_mode(double alive_ratio) {
+        if (alive_ratio <= stop_evict_ratio) {
+          eviction_mode = eviction_mode_t::STOP;
+        } else if (alive_ratio <= default_evict_ratio) {
+          if (eviction_mode > eviction_mode_t::DEFAULT) {
+            eviction_mode = eviction_mode_t::DEFAULT;
+          }
+        } else if (alive_ratio <= fast_evict_ratio) {
+          if (eviction_mode < eviction_mode_t::DEFAULT) {
+            eviction_mode = eviction_mode_t::DEFAULT;
+          }
+        } else {
+          assert(alive_ratio > fast_evict_ratio);
+          eviction_mode = eviction_mode_t::FAST;
+        }
+      }
+    };
+
+    bool should_do_fast_eviction() const {
```

Comment:
> This is not reasonable to me, currently. I think there is an implicit invariant from `AsyncCleaner` that if `should_clean_space()` is false, it doesn't expect `clean_space()` to be called (worth to supplement the interface comments).
> 
> The issue of the current `should_clean_space()` is that it is a mixture of both "can clean space" and "request to clean space". We need to adjust the Cleaner before implementing the "force clean" feature.

You are right, it is part of the "force clean" feature which is to solve the problem mentioned in https://github.com/ceph/ceph/pull/48691#discussion_r1032473643.

> I think it is not a blocking issue for this PR and we can leave it in a followup PR with a TODO to remember, if need further discussions.

such an issue is rarely seen in more realistic scenes, I'll remove it in this PR.

Reply:
This problem occurs in this situation:
1. Seastore has multiple tiers;
2. Given an rbd image and write data to it sequentially;
3. At some points, projected_avail_ratio=0.100315, reclaim_ratio=0.0997815(from https://github.com/ceph/ceph/pull/48691#discussion_r1032473643), invoking `AsyncCleaner::should_clean_space` will return false;
4. Then there is a client IO transaction trying to reserve space on the main cleaner. The reservation will fail because the new projected_avail_ratio < 0.1 (available_ratio_hard_limit), but the cleaner won't start.

I think this is the only scenario that the "force clean" that will make sense.

The workaround in my mind contains:
1. loose the restriction of start condition:
```cpp
bool should_clean_space() const final { [2/3]
    assert(background_callback->is_ready());
    if (get_segments_reclaimable() == 0) {
      return false;
    }
    auto aratio = segments.get_available_ratio();
    auto rratio = get_reclaim_ratio();
    return (
      (aratio < config.available_ratio_hard_limit +
       1.0 / segments.get_num_segments()) ||
      ((aratio < config.available_ratio_gc_max) &&
       (rratio > config.reclaim_ratio_gc_threshold))
    );
  }
```
2. something like `should_do_fast_eviction` to ensure the cleaner will start even if `should_clean_space` returns false.

Do you have any other suggestions?

Reply:
I see the problem.

I think the only restriction of the start condition is that there must be segments available to reclaim.

How about this?

```c++
// (called by both background_should_run() and do_background_cycle())
bool BackgroundProcess::should_clean_main() {
  return (
    main_cleaner->should_clean_space() ||
    (main_cleaner->can_clean_space() &&
     eviction_state.is_mode_fast()) // true implies has_cold_tier()
  );
}

SegmentCleaner::clean_space_ret
SegmentCleaner::clean_space()
{
  LOG_PREFIX(SegmentCleaner::clean_space);
  assert(background_callback->is_ready());
  assert(can_clean_space());
  < ... left unchanged ... >
}

bool SegmentCleaner::should_clean_space() const final {
  assert(background_callback->is_ready());
  if (!can_clean_space()) {
    return false;
  }
  auto aratio = segments.get_available_ratio();
  auto rratio = get_reclaim_ratio();
  return (
    (aratio < config.available_ratio_hard_limit +
     1.0 / segments.get_num_segments()) ||
    ((aratio < config.available_ratio_gc_max) &&
     (rratio > config.reclaim_ratio_gc_threshold))
  );
}
  
bool SegmentCleaner::can_clean_space() const final {
  assert(background_callback->is_ready());
  if (get_segments_reclaimable() == 0) {
    return false;
  } else {
    return true;
  }
}
```

Reply:
After rethinking this problem, I think loosing the restriction of the start condition (the `config.available_ratio_hard_limit +1.0 / segments.get_num_segments()` part) is not a good idea.

If we do so, it causes a new problem that when seastore only has one tier, at such case the cleaner process will keep working even if there is little space to reclaim.

Using `eviction_state.is_mode_fast()` to start cleaner is enough to solve this problem.